File: src/org/vaadin/multiprobe/data/ProbeData.java
package org.vaadin.multiprobe.data;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.Writer;
import java.io.IOException;
import java.io.Serializable;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Collection;
import java.util.Date;
import java.util.HashMap;


public class ProbeData implements Serializable {
    private static final long serialVersionUID = 3502338010463752172L;

    static final SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.S");
    
    HashMap<Integer, Probe> probes = new HashMap<Integer, Probe>(); 
    
    BufferedWriter out;

    static ProbeData instance = new ProbeData();
    
    public static ProbeData get() {
        return instance;
    }

    private ProbeData() {
        String datafile = System.getenv("HOME") + "/analprobe.dat";
        System.out.println("Opening data journal " + datafile);
        
        // Read data collected so far, but no data older than a day
        try {
            FileReader reader = new FileReader(datafile);
            BufferedReader in = new BufferedReader(reader);
            int readLines = 0;
            for (String line = in.readLine(); line != null; line = in.readLine()) {
                line = line.trim();
                String split[] = line.split(" ");
                if (split.length == 3)
                    try {
                        Date date = dateFormat.parse(split[0]);
                        int probeId = Integer.parseInt(split[1]);
                        int value = Integer.parseInt(split[2]);
                        getProbe(probeId).add(new ProbeEntry(date, value), null);
                    } catch (ParseException e) {
                        System.err.println("Invalid format '" +
                            e.getMessage() + " in " + line);
                    }
                else
                    System.err.println("Invalid format: " + line);
            }
            in.close();
            System.out.println(readLines + " entries read");
        } catch (FileNotFoundException e) {
            System.out.println("Journal not found, creating one.");
        } catch (IOException e) {
            System.err.println(e.getMessage());
        }
        
        // Start writing more
        try {
            FileWriter writer = new FileWriter(datafile, true);
            out = new BufferedWriter(writer);
        } catch (IOException e) {
            System.err.println(e.getMessage());
        }
    }
    
    public Collection<Probe> getProbes() {
        return probes.values();
    }
    
    public Probe getProbe(int probeId) {
        if (! probes.containsKey(probeId))
            probes.put(probeId, new Probe(this, probeId));
        return probes.get(probeId);
    }
    
    public Writer getJournalWriter() {
        return out;
    }
}

File: src/org/vaadin/multiprobe/data/ProbeData.scala
package org.vaadin.multiprobe.data

import java.text.SimpleDateFormat
import scala.collection.mutable.HashMap
import java.io.BufferedWriter
import java.io.FileNotFoundException
import java.io.IOException
import java.io.BufferedReader
import java.io.FileReader
import scala.io.Source
import java.util.Date
import java.text.ParseException
import java.io.FileWriter

/**
 * Contains the data from all probes.
 *
 * This is a singleton object, to which the collector writes data and UIs read from.
 */
object ProbeData {
  val dateFormat: SimpleDateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.S");
  var probes: HashMap[Int, Probe] = new HashMap[Int, Probe]
  var journal: BufferedWriter = _

  {
    val datafile: String = System.getenv("HOME") + "/analprobe.dat"
    println("Opening data journal " + datafile)

    try {
      val reader: FileReader = new FileReader(datafile)
      val in: BufferedReader = new BufferedReader(reader)
      var readLines: Int = 0

      for (line: String <- Source.fromFile(datafile).getLines) {
        val trimmed: String = line.trim()
        val split = trimmed.split(" ")
        if (split.length == 3) {
          try {
            // Probe events include a date, probe ID, and measurement
            val date: Date = dateFormat.parse(split(0))
            val probeId: Int = Integer.parseInt(split(1))
            val value: Int = Integer.parseInt(split(2))

            getOrCreateProbe(probeId).add(new ProbeEntry(date, value), null)
          } catch {
            case e: ParseException =>
              System.err.println("Invalid format '" +
                e.getMessage() + " in " + line)
          }
        } else
          System.err.println("Invalid format: " + line)
      }

      in.close()
      println(readLines + " entries read")
    } catch {
      case e: FileNotFoundException =>
        println("Journal not found, creating one.")
      case e: IOException =>
        System.err.println(e.getMessage())
    }

    // Start writing more
    try {
      val writer: FileWriter = new FileWriter(datafile, true);
      journal = new BufferedWriter(writer);
    } catch {
      case e: IOException =>
        System.err.println(e.getMessage());
    }
  }

  def getOrCreateProbe(probeId: Int): Probe = {
    if (probes.contains(probeId))
      probes(probeId)
    else {
      val newProbe = new Probe(probes.size)
      probes.put(newProbe.id, newProbe)
      newProbe
    }
  }
}
